cpp/src/parquet/column_reader.cc

```diff
@@ -1401,26 +1423,22 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
         int16_t* def_levels = this->def_levels() + levels_written_;
         int16_t* rep_levels = this->rep_levels() + levels_written_;
 
-        // Not present for non-repeated fields
-        int64_t levels_read = 0;
+        int64_t levels_read = this->ReadDefinitionLevels(batch_size, def_levels);
         if (this->max_rep_level_ > 0) {
-          levels_read = this->ReadDefinitionLevels(batch_size, def_levels);
-          if (this->ReadRepetitionLevels(batch_size, rep_levels) != levels_read) {
-            throw ParquetException("Number of decoded rep / def levels did not match");
+          int64_t rep_levels_read = this->ReadRepetitionLevels(batch_size, rep_levels);
+          if (rep_levels_read != levels_read) {
+            throw ParquetException(kErrorRepDefLevelNotMatchesNumValues);
           }
-        } else if (this->max_def_level_ > 0) {
-          levels_read = this->ReadDefinitionLevels(batch_size, def_levels);
         }
 
-        // Exhausted column chunk
-        if (levels_read == 0) {
-          break;
+        if (ARROW_PREDICT_FALSE(batch_size != levels_read)) {
+          throw ParquetException(kErrorRepDefLevelNotMatchesNumValues);
         }
 
         levels_written_ += levels_read;
```

Comment:
Can also it be `levels_written_ += batch_size`? i.e. we don't need `levels_read` at all.